src/DataDictionary/graph/GraphCalculator/graphCalculatorHelper.js

```diff
@@ -0,0 +1,520 @@
+            Object.keys(originNode.properties).length - requiredPropertiesCount : 0;
+          let nodeLevel = 0;
+          if (originNode && originNode.positionIndex && originNode.positionIndex.length >= 2) {
+            nodeLevel = originNode.positionIndex[1];
+          }
+          return {
+            id: n.name,
+            type: n.type,
+            boundingBox,
+            topCenterX,
+            topCenterY,
+            width,
+            height: rectHeight,
+            color: nodeColor,
+            iconRadius,
+            textPadding,
+            fontSize,
+            textLineGap,
+            names: nodeNames,
+            label: n.label,
+            level: nodeLevel,
+            outLinks,
+            inLinks,
+            _gvid: n._gvid,
+            requiredPropertiesCount,
+            optionalPropertiesCount,
+          };
+        });
+
+      // draw edges
+      const renderedEdges = renderedJSON.edges
+        .map((edge) => {
+          const controlPoints = edge._draw_[1].points;
+          let pathString = `M${controlPoints[0].join(',')}C${controlPoints.slice(1)
+            .map(pair => `${pair[0]},${pair[1]}`).join(' ')}`;
+          const sourceNode = renderedNodes.find(node => node._gvid === edge.tail);
+          const targetNode = renderedNodes.find(node => node._gvid === edge.head);
+          if (sourceNode.level === targetNode.level + 1) {
+            const sourePosition = [
+              (sourceNode.boundingBox.x1 + sourceNode.boundingBox.x2) / 2,
+              sourceNode.boundingBox.y1,
+            ];
+            const targetPosition = [
+              (targetNode.boundingBox.x1 + targetNode.boundingBox.x2) / 2,
+              targetNode.boundingBox.y2,
+            ];
+            pathString = `M${sourePosition[0]} ${sourePosition[1]} 
+              L ${targetPosition[0]} ${targetPosition[1]}`;
+          }
+          const required = edges
+            .find(e => (e.source.id === sourceNode.id && e.target.id === targetNode.id))
+            .required;
+          return {
+            source: sourceNode.id,
+            target: targetNode.id,
+            controlPoints,
+            pathString,
+            required,
+          };
+        });
+
+      // get bounding box for whole graph
+      const graphBoundingBox = renderedJSON._draw_.find(entry => entry.op === 'P').points;
+
+      const layoutResult = {
+        nodes: renderedNodes,
+        edges: renderedEdges,
+        graphBoundingBox,
+      };
+      return layoutResult;
+    })
+    .catch((e) => {
+      throw e;
+    });
+};
+/* eslint-enable no-underscore-dangle */
+
+export const getNodeIDsThatHaveNoInLinks = (
+  subgraphNodeIDs,
+  subgraphEdges,
+  wholeGraphNodes,
+) => {
+  const resultIDs = [];
+  subgraphNodeIDs.forEach((nodeID) => {
+    const node = wholeGraphNodes.find(n => n.id === nodeID);
+    const inLinks = node.inLinks
+      .filter(inNodeID => subgraphNodeIDs.includes(inNodeID))
```

Comment:
this logic and the later one in `getNodeIDsThatHaveNoOutLinks` are very similar. Should we:
- create a filter function receive lists as parameters
- for filter that are different you can pass it as a parameter
- and give the filter functions meaningful names?